File: ontrack-extension-api/src/test/java/net/nemerosa/ontrack/extension/api/support/TestSimplePropertyType.java
package net.nemerosa.ontrack.extension.api.support;

import com.fasterxml.jackson.databind.JsonNode;
import net.nemerosa.ontrack.extension.support.AbstractPropertyType;
import net.nemerosa.ontrack.model.form.Form;
import net.nemerosa.ontrack.model.form.Text;
import net.nemerosa.ontrack.model.security.ProjectEdit;
import net.nemerosa.ontrack.model.security.SecurityService;
import net.nemerosa.ontrack.model.structure.ProjectEntity;
import net.nemerosa.ontrack.model.structure.ProjectEntityType;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.util.EnumSet;
import java.util.Set;
import java.util.function.Function;

@Component
public class TestSimplePropertyType extends AbstractPropertyType<TestSimpleProperty> {

    @Autowired
    public TestSimplePropertyType(TestExtensionFeature extensionFeature) {
        super(extensionFeature);
    }

    @Override
    public String getName() {
        return "Simple value";
    }

    @Override
    public String getDescription() {
        return "Value.";
    }

    @Override
    public Set<ProjectEntityType> getSupportedEntityTypes() {
        return EnumSet.allOf(ProjectEntityType.class);
    }

    @Override
    public boolean canEdit(ProjectEntity entity, SecurityService securityService) {
        return securityService.isProjectFunctionGranted(entity, ProjectEdit.class);
    }

    @Override
    public boolean canView(ProjectEntity entity, SecurityService securityService) {
        return true;
    }

    @Override
    public Form getEditionForm(ProjectEntity entity, TestSimpleProperty value) {
        return Form.create()
                .with(
                        Text.of("value")
                                .label("Value")
                                .value(value != null ? value.getValue() : "")
                )
                ;
    }

    @Override
    public TestSimpleProperty fromClient(JsonNode node) {
        return fromStorage(node);
    }

    @Override
    public TestSimpleProperty fromStorage(JsonNode node) {
        return parse(node, TestSimpleProperty.class);
    }

    @Override
    public String getSearchKey(TestSimpleProperty value) {
        return value.getValue();
    }

    @Override
    public TestSimpleProperty replaceValue(TestSimpleProperty value, Function<String, String> replacementFunction) {
        return new TestSimpleProperty(
                replacementFunction.apply(value.getValue())
        );
    }
}

File: ontrack-extension-api/src/testFixtures/kotlin/net/nemerosa/ontrack/extension/api/support/TestSimplePropertyType.kt
package net.nemerosa.ontrack.extension.api.support

import com.fasterxml.jackson.databind.JsonNode
import net.nemerosa.ontrack.extension.support.AbstractPropertyType
import net.nemerosa.ontrack.model.json.schema.JsonType
import net.nemerosa.ontrack.model.json.schema.JsonTypeBuilder
import net.nemerosa.ontrack.model.json.schema.toType
import net.nemerosa.ontrack.model.security.ProjectEdit
import net.nemerosa.ontrack.model.security.SecurityService
import net.nemerosa.ontrack.model.structure.ProjectEntity
import net.nemerosa.ontrack.model.structure.ProjectEntityType
import net.nemerosa.ontrack.model.structure.PropertySearchArguments
import org.springframework.stereotype.Component
import java.util.*

@Component
class TestSimplePropertyType(
    extensionFeature: TestExtensionFeature
) : AbstractPropertyType<TestSimpleProperty>(extensionFeature) {

    override val name: String = "Simple value"

    override val description: String = "Value."

    override val supportedEntityTypes: Set<ProjectEntityType> = EnumSet.allOf(ProjectEntityType::class.java)

    override fun createConfigJsonType(jsonTypeBuilder: JsonTypeBuilder): JsonType =
        jsonTypeBuilder.toType(TestSimpleProperty::class)

    override fun canEdit(entity: ProjectEntity, securityService: SecurityService): Boolean =
        securityService.isProjectFunctionGranted(entity, ProjectEdit::class.java)

    override fun canView(entity: ProjectEntity, securityService: SecurityService): Boolean = true

    override fun fromClient(node: JsonNode): TestSimpleProperty = fromStorage(node)

    override fun fromStorage(node: JsonNode): TestSimpleProperty = parse(node, TestSimpleProperty::class)

    override fun replaceValue(
        value: TestSimpleProperty,
        replacementFunction: (String) -> String
    ): TestSimpleProperty {
        return TestSimpleProperty(
            replacementFunction(value.value)
        )
    }

    override fun getSearchArguments(token: String): PropertySearchArguments? {
        return PropertySearchArguments(
            null,
            "pp.json->>'value' like :value",
            Collections.singletonMap("value", "%$token%")
        )
    }

    override fun containsValue(value: TestSimpleProperty, propertyValue: String): Boolean =
        value.value.contains(propertyValue, ignoreCase = true)
}
